File: dotnet/DesignPatterns/StructuralDesignPatterns/ProxyDesignPattern/ProxyDesignPattern.SmartReferenceProxyExample/DatabaseConnectionProxy.cs
namespace ProxyDesignPattern.SmartReferenceProxyExample;

/// <summary>
/// Smart Reference Proxy - Tracks references and adds extra behavior:
/// - Reference counting
/// - Access logging
/// - Last access time tracking
/// - Auto-close when reference count reaches zero
/// </summary>
public class DatabaseConnectionProxy : IDatabaseConnection
{
    private readonly RealDatabaseConnection _realConnection;
    private readonly string _proxyId;
    
    private int _referenceCount;
    private int _queryCount;
    private DateTime _lastAccessTime;
    private bool _isClosed;

    public DatabaseConnectionProxy()
    {
        _proxyId = Guid.NewGuid().ToString()[..4].ToUpper();
        _realConnection = new RealDatabaseConnection();
        _referenceCount = 1;
        _lastAccessTime = DateTime.Now;
        
        Console.WriteLine($"[Proxy-{_proxyId}] Smart proxy created. Reference count: {_referenceCount}");
    }

    public void AddReference()
    {
        _referenceCount++;
        Console.WriteLine($"[Proxy-{_proxyId}] Reference added. Count: {_referenceCount}");
    }

    public void ReleaseReference()
    {
        _referenceCount--;
        Console.WriteLine($"[Proxy-{_proxyId}] Reference released. Count: {_referenceCount}");

        if (_referenceCount <= 0)
        {
            Console.WriteLine($"[Proxy-{_proxyId}] No more references - auto-closing connection");
            Close();
        }
    }

    public void ExecuteQuery(string query)
    {
        if (_isClosed)
            throw new InvalidOperationException("Connection is closed");

        _queryCount++;
        _lastAccessTime = DateTime.Now;

        // Log access
        Console.WriteLine($"[Proxy-{_proxyId}] Query #{_queryCount} at {_lastAccessTime:HH:mm:ss}");
        
        _realConnection.ExecuteQuery(query);
    }

    public void Close()
    {
        if (!_isClosed)
        {
            _isClosed = true;
            _realConnection.Close();
            Console.WriteLine($"[Proxy-{_proxyId}] Statistics: {_queryCount} queries executed");
        }
    }

    public void PrintStatistics()
    {
        Console.WriteLine($"\n[Proxy-{_proxyId}] --- Statistics ---");
        Console.WriteLine($"[Proxy-{_proxyId}] Reference Count: {_referenceCount}");
        Console.WriteLine($"[Proxy-{_proxyId}] Queries Executed: {_queryCount}");
        Console.WriteLine($"[Proxy-{_proxyId}] Last Access: {_lastAccessTime:HH:mm:ss}");
        Console.WriteLine($"[Proxy-{_proxyId}] Is Closed: {_isClosed}");
    }
}

File: nodejs/DesignPatterns/StructuralDesignPatterns/ProxyDesignPattern/ProxyDesignPattern.SmartReferenceProxyExample/src/DatabaseConnectionProxy.ts
import { IDatabaseConnection } from "./IDatabaseConnection";
import { RealDatabaseConnection } from "./RealDatabaseConnection";

/**
 * Smart Reference Proxy - Tracks references and adds extra behavior:
 * - Reference counting
 * - Access logging
 * - Last access time tracking
 * - Auto-close when reference count reaches zero
 */
export class DatabaseConnectionProxy implements IDatabaseConnection {
  private readonly _realConnection: RealDatabaseConnection;
  private readonly _proxyId: string;

  private _referenceCount: number;
  private _queryCount: number;
  private _lastAccessTime: Date;
  private _isClosed: boolean;

  constructor() {
    this._proxyId = Math.random().toString(36).substring(2, 6).toUpperCase();
    this._realConnection = new RealDatabaseConnection();
    this._referenceCount = 1;
    this._queryCount = 0;
    this._lastAccessTime = new Date();
    this._isClosed = false;

    console.log(
      `[Proxy-${this._proxyId}] Smart proxy created. Reference count: ${this._referenceCount}`
    );
  }

  addReference(): void {
    this._referenceCount++;
    console.log(`[Proxy-${this._proxyId}] Reference added. Count: ${this._referenceCount}`);
  }

  releaseReference(): void {
    this._referenceCount--;
    console.log(`[Proxy-${this._proxyId}] Reference released. Count: ${this._referenceCount}`);

    if (this._referenceCount <= 0) {
      console.log(`[Proxy-${this._proxyId}] No more references - auto-closing connection`);
      this.close();
    }
  }

  executeQuery(query: string): void {
    if (this._isClosed) {
      throw new Error("Connection is closed");
    }

    this._queryCount++;
    this._lastAccessTime = new Date();

    const timeStr = this._lastAccessTime.toTimeString().slice(0, 8);
    console.log(`[Proxy-${this._proxyId}] Query #${this._queryCount} at ${timeStr}`);

    this._realConnection.executeQuery(query);
  }

  close(): void {
    if (!this._isClosed) {
      this._isClosed = true;
      this._realConnection.close();
      console.log(`[Proxy-${this._proxyId}] Statistics: ${this._queryCount} queries executed`);
    }
  }

  printStatistics(): void {
    const timeStr = this._lastAccessTime.toTimeString().slice(0, 8);
    console.log(`\n[Proxy-${this._proxyId}] --- Statistics ---`);
    console.log(`[Proxy-${this._proxyId}] Reference Count: ${this._referenceCount}`);
    console.log(`[Proxy-${this._proxyId}] Queries Executed: ${this._queryCount}`);
    console.log(`[Proxy-${this._proxyId}] Last Access: ${timeStr}`);
    console.log(`[Proxy-${this._proxyId}] Is Closed: ${this._isClosed}`);
  }
}
